src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -382,39 +384,178 @@ See Zulip thread:
+
+/-- The basis $\{1, Y\}$ for the coordinate ring $R[W]$ over the polynomial ring $R[X]$.
+
+Given a Weierstrass curve `W`, write `W^.coordinate_ring.basis` for this basis. -/
+protected noncomputable def basis : basis (fin 2) R[X] W.coordinate_ring :=
+(subsingleton_or_nontrivial R).by_cases (λ _, by exactI default) $ λ _, by exactI
+  (basis.reindex (adjoin_root.power_basis' W.monic_polynomial).basis $
+    fin_congr $ W.nat_degree_polynomial)
+
+lemma basis_apply [nontrivial R] (n : fin 2) :
+  W^.coordinate_ring.basis n
+    = (adjoin_root.power_basis' W.monic_polynomial).gen
+      ^ ((fin_congr W.nat_degree_polynomial).symm n : ℕ) :=
+by rw [coordinate_ring.basis, or.by_cases, dif_neg $ λ h : subsingleton R,
+         by exactI not_subsingleton R h, basis.reindex_apply, power_basis.basis_eq_pow]
+
+lemma basis_zero : W^.coordinate_ring.basis 0 = 1 :=
+by { nontriviality R, simpa only [basis_apply] using pow_zero _ }
+
+lemma basis_one : W^.coordinate_ring.basis 1 = adjoin_root.mk W.polynomial X :=
+by { nontriviality R, simpa only [basis_apply] using pow_one _ }
+
+@[simp] lemma coe_basis :
+  (W^.coordinate_ring.basis : fin 2 → W.coordinate_ring) = ![1, adjoin_root.mk W.polynomial X] :=
+by { ext n, fin_cases n, exacts [basis_zero W, basis_one W] }
+
+variable {W}
+
+lemma smul (x : R[X]) (y : W.coordinate_ring) : x • y = adjoin_root.mk W.polynomial (C x) * y :=
+(algebra_map_smul W.coordinate_ring x y).symm
+
+lemma smul_basis_eq_zero {p q : R[X]}
+  (hpq : p • 1 + q • adjoin_root.mk W.polynomial X = 0) : p = 0 ∧ q = 0 :=
+begin
+  have h := fintype.linear_independent_iff.mp (coordinate_ring.basis W).linear_independent ![p, q],
+  erw [fin.sum_univ_succ, basis_zero, fin.sum_univ_one, basis_one] at h,
+  exact ⟨h hpq 0, h hpq 1⟩
+end
+
+lemma exists_smul_basis_eq (x : W.coordinate_ring) :
+  ∃ p q : R[X], p • 1 + q • adjoin_root.mk W.polynomial X = x :=
+begin
+  have h := (coordinate_ring.basis W).sum_equiv_fun x,
+  erw [fin.sum_univ_succ, fin.sum_univ_one, basis_zero, basis_one] at h,
+  exact ⟨_, _, h⟩
+end
+
+variable (W)
+
+lemma smul_basis_mul_C (p q : R[X]) :
+  (p • 1 + q • adjoin_root.mk W.polynomial X) * adjoin_root.mk W.polynomial (C y)
+    = ((p * y) • 1 + (q * y) • adjoin_root.mk W.polynomial X) :=
+by { simp only [smul, map_mul], ring1 }
+
+lemma smul_basis_mul_Y (p q : R[X]) :
+  (p • 1 + q • adjoin_root.mk W.polynomial X) * adjoin_root.mk W.polynomial X
+    = (q * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)) • 1
+      + (p - q * (C W.a₁ * X + C W.a₃)) • adjoin_root.mk W.polynomial X :=
+begin
+  have Y_sq : adjoin_root.mk W.polynomial X ^ 2 = adjoin_root.mk W.polynomial
+    (C (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆) - C (C W.a₁ * X + C W.a₃) * X) :=
+  adjoin_root.mk_eq_mk.mpr ⟨1, by { simp only [weierstrass_curve.polynomial], ring1 }⟩,
+  simp only [smul, add_mul, mul_assoc, ← sq, Y_sq, map_sub, map_mul],
+  ring1
+end
+
+/-! ### Norms on the coordinate ring -/
```

Comment:
This starts to be mathematically nontrivial. Can you add a short explanation of what you are doing? A general explanation or some docstrings, or maybe both.